rust/src/operations/update.rs

```diff
@@ -0,0 +1,921 @@
+            let table = DeltaTable::new_with_state(this.object_store, this.snapshot);
+
+            Ok((table, metrics))
+        })
+    }
+}
+
+#[derive(Debug)]
+struct UpdateCountExec {
+    parent: Arc<dyn ExecutionPlan>,
+    metrics: ExecutionPlanMetricsSet,
+}
+
+impl UpdateCountExec {
+    pub fn new(parent: Arc<dyn ExecutionPlan>) -> Self {
+        UpdateCountExec {
+            parent,
+            metrics: ExecutionPlanMetricsSet::new(),
+        }
+    }
+}
+
+impl ExecutionPlan for UpdateCountExec {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn schema(&self) -> arrow_schema::SchemaRef {
+        self.parent.schema()
+    }
+
+    fn output_partitioning(&self) -> datafusion::physical_plan::Partitioning {
+        self.parent.output_partitioning()
+    }
+
+    fn output_ordering(&self) -> Option<&[datafusion_physical_expr::PhysicalSortExpr]> {
+        self.parent.output_ordering()
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![self.parent.clone()]
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<datafusion::execution::context::TaskContext>,
+    ) -> datafusion_common::Result<datafusion::physical_plan::SendableRecordBatchStream> {
+        let res = self.parent.execute(partition, context)?;
+        Ok(Box::pin(UpdateCountStream {
+            schema: self.schema(),
+            input: res,
+            metrics: self.metrics.clone(),
+        }))
+    }
+
+    fn statistics(&self) -> datafusion_common::Statistics {
+        self.parent.statistics()
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> datafusion_common::Result<Arc<dyn ExecutionPlan>> {
+        ExecutionPlan::with_new_children(self.parent.clone(), children)
+    }
+
+    fn metrics(&self) -> Option<MetricsSet> {
+        Some(self.metrics.clone_inner())
+    }
+}
+
+struct UpdateCountStream {
+    schema: SchemaRef,
+    input: SendableRecordBatchStream,
+    metrics: ExecutionPlanMetricsSet,
+}
+
+impl Stream for UpdateCountStream {
+    type Item = DataFusionResult<RecordBatch>;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        self.input.poll_next_unpin(cx).map(|x| match x {
+            Some(Ok(batch)) => {
+                let array = batch.column_by_name("__delta_rs_update_predicate").unwrap();
+                let copied_rows = array.null_count();
+                let num_updated = array.len() - copied_rows;
+                let c1 = MetricBuilder::new(&self.metrics).global_counter("num_updated_rows");
+                c1.add(num_updated);
+
+                let c2 = MetricBuilder::new(&self.metrics).global_counter("num_copied_rows");
+                c2.add(copied_rows);
+                Some(Ok(batch))
```

Comment:
This is cool!

Reply:
Indeed it is!

Reply:
Thanks it took me a bit to arrive to this solution but I'm glad ended up this simple.